File: ios/SnapSell/UI/Screens/Review/ReviewViewModel.swift
import CoreGraphics
import Foundation
import Observation
import UIKit

/// Runs segmentation on the current photo, tracks which subjects the user
/// wants, and on confirm renders cutouts + inserts placeholder items that the
/// Confirm screen then identifies.
@MainActor
@Observable
final class ReviewViewModel {
    static let minManualSize: CGFloat = 24

    private(set) var image: UIImage?
    private(set) var segments: [Segment] = []
    private(set) var selected: Set<Int> = []
    private(set) var segmenting = false
    private(set) var committing = false
    var manualMode = false
    /// Vision could not run: boxes are the only way in.
    private(set) var outlinerUnavailable = false
    private(set) var error: String?
    /// Cutouts kept from earlier photos in this capture session.
    private(set) var pendingCount = 0

    var selectedCount: Int { selected.count + pendingCount }

    @ObservationIgnored private let app: AppContainer
    private var session: CaptureSession { app.captureSession }

    init(app: AppContainer) {
        self.app = app
        pendingCount = session.pendingCutouts.count
        guard let photo = session.current else {
            error = "No photo to review. Go back and take one."
            return
        }
        image = photo.workImage
        segments = photo.segments
        selected = photo.selected
        if photo.segments.isEmpty { segment(photo) }
    }

    private func segment(_ photo: CaptureSession.Photo) {
        segmenting = true
        error = nil
        Task {
            do {
                let found = try await app.segmenter.segment(photo.workImage).segments
                photo.segments = found
                // Everything found is selected by default; the user deselects.
                photo.selected = Set(found.map(\.index))
                segments = found
                selected = photo.selected
                manualMode = found.isEmpty
                error = found.isEmpty ? "No objects found. Draw a box around each thing you want to sell." : nil
            } catch {
                manualMode = true
                outlinerUnavailable = true
            }
            segmenting = false
        }
    }

    func toggle(_ index: Int) {
        guard let photo = session.current else { return }
        if photo.selected.contains(index) { photo.selected.remove(index) } else { photo.selected.insert(index) }
        selected = photo.selected
    }

    /// Toggle whichever segment contains the tapped point (smallest wins).
    func tap(at p: CGPoint) {
        guard let hit = segments.filter({ $0.contains(p.x, p.y) }).min(by: { $0.area < $1.area }) else { return }
        toggle(hit.index)
    }

    /// Adds a user-drawn rectangle (photo pixel coordinates) as a selected segment.
    func addManualRect(_ rect: CGRect) {
        guard let photo = session.current else { return }
        let bounds = CGRect(origin: .zero, size: photo.workImage.size)
        let clipped = rect.standardized.intersection(bounds)
        guard !clipped.isNull, clipped.width >= Self.minManualSize, clipped.height >= Self.minManualSize else { return }
        let index = (photo.segments.map(\.index).max() ?? -1) + 1
        photo.segments.append(Segment.manual(index: index, rect: clipped))
        photo.selected.insert(index)
        segments = photo.segments
        selected = photo.selected
        manualMode = false
        error = nil
    }

    func remove(_ index: Int) {
        guard let photo = session.current else { return }
        photo.segments.removeAll { $0.index == index }
        photo.selected.remove(index)
        segments = photo.segments
        selected = photo.selected
    }

    /// "Retake" for one item: drop that segment, stash every other selected
    /// cutout from this photo, then let the caller open the camera again.
    func retake(_ index: Int, onReady: @escaping () -> Void) {
        remove(index)
        Task {
            try? await stashSelected()
            onReady()
        }
    }

    private func stashSelected() async throws {
        guard let photo = session.current else { return }
        let toRender = photo.segments.filter { photo.selected.contains($0.index) }
        if toRender.isEmpty { return }
        let store = app.photoStore
        let work = photo.workImage
        let file = photo.file
        let cutouts: [CaptureSession.Cutout] = try await Task.detached(priority: .userInitiated) {
            try toRender.map { seg in
                guard let img = CutoutRenderer.render(work, seg) else { throw AppError("Couldn't render a cutout.") }
                return CaptureSession.Cutout(file: try store.saveCutout(img), originalPhoto: file)
            }
        }.value
        session.pendingCutouts.append(contentsOf: cutouts)
        photo.selected = []
        selected = []
        pendingCount = session.pendingCutouts.count
    }

    /// Render every selected cutout, insert placeholder items, hand their ids to Confirm.
    func identify(onItems: @escaping ([String]) -> Void) {
        guard !committing else { return }
        committing = true
        error = nil
        Task {
            do {
                try await stashSelected()
                let cutouts = session.pendingCutouts
                if cutouts.isEmpty {
                    committing = false
                    error = "Select at least one item."
                    return
                }
                let now = app.clock.nowMillis()
                let items = cutouts.map { cutout in
                    Item(
                        id: Ids.newId(),
                        // Blank name marks "not identified yet"; ConfirmViewModel fills it in.
                        name: "", category: "", condition: .good, searchQuery: "", confidence: 0,
                        photoFile: cutout.file, cutoutFile: cutout.file, originalPhotoFile: cutout.originalPhoto,
                        createdAt: now, updatedAt: now
                    )
                }
                try app.inventory.saveItems(items)
                session.clear()
                onItems(items.map(\.id))
            } catch {
                committing = false
                self.error = "Could not prepare cutouts: \(AppError.describe(error).message)"
            }
        }
    }
}

File: android/app/src/main/java/com/alexivanov/snapsell/ui/review/ReviewViewModel.kt
package com.alexivanov.snapsell.ui.review

import android.graphics.Bitmap
import android.graphics.Rect
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.alexivanov.snapsell.AppContainer
import com.alexivanov.snapsell.core.Ids
import com.alexivanov.snapsell.data.local.ItemEntity
import com.alexivanov.snapsell.domain.Condition
import com.alexivanov.snapsell.vision.CaptureSession
import com.alexivanov.snapsell.vision.Segment
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

data class ReviewUiState(
    val bitmap: Bitmap? = null,
    val segments: List<Segment> = emptyList(),
    val selected: Set<Int> = emptySet(),
    val segmenting: Boolean = false,
    val committing: Boolean = false,
    val manualMode: Boolean = false,
    /** ML Kit could not run (model missing, emulator GL, ...): boxes are the only way in. */
    val outlinerUnavailable: Boolean = false,
    val error: String? = null,
    /** Cutouts kept from earlier photos in this capture session. */
    val pendingCount: Int = 0,
) {
    val selectedCount: Int get() = selected.size + pendingCount
}

/**
 * Runs segmentation on the current photo, tracks which subjects the user
 * wants, and on confirm renders cutouts + inserts placeholder items that the
 * Confirm screen then identifies.
 */
class ReviewViewModel(private val container: AppContainer) : ViewModel() {
    private val session: CaptureSession get() = container.captureSession

    private val _state = MutableStateFlow(ReviewUiState(pendingCount = session.pendingCutouts.size))
    val state: StateFlow<ReviewUiState> = _state

    init {
        val photo = session.current
        if (photo == null) {
            _state.update { it.copy(error = "No photo to review. Go back and take one.") }
        } else {
            _state.update { it.copy(bitmap = photo.workBitmap, segments = photo.segments, selected = photo.selected.toSet()) }
            if (photo.segments.isEmpty()) segment(photo)
        }
    }

    private fun segment(photo: CaptureSession.Photo) {
        _state.update { it.copy(segmenting = true, error = null) }
        viewModelScope.launch {
            val result = runCatching { container.segmenter.segment(photo.workBitmap) }
            result.onSuccess { segments ->
                photo.segments = segments
                // Everything ML Kit found is selected by default; the user deselects.
                photo.selected.clear()
                photo.selected.addAll(segments.map { it.index })
                _state.update {
                    it.copy(
                        segmenting = false,
                        segments = segments,
                        selected = photo.selected.toSet(),
                        manualMode = segments.isEmpty(),
                        error = if (segments.isEmpty()) "No objects found. Draw a box around each thing you want to sell." else null,
                    )
                }
            }.onFailure { e ->
                // Typical cause: the segmentation model has not been downloaded by Play services yet.
                _state.update {
                    it.copy(segmenting = false, manualMode = true, outlinerUnavailable = true, error = null)
                }
            }
        }
    }

    fun toggle(index: Int) {
        val photo = session.current ?: return
        if (!photo.selected.remove(index)) photo.selected.add(index)
        _state.update { it.copy(selected = photo.selected.toSet()) }
    }

    /** Toggle whichever segment contains the tapped bitmap point (smallest wins). */
    fun tapAt(x: Int, y: Int) {
        val hit = _state.value.segments
            .filter { it.contains(x, y) }
            .minByOrNull { it.width.toLong() * it.height }
            ?: return
        toggle(hit.index)
    }

    fun setManualMode(enabled: Boolean) = _state.update { it.copy(manualMode = enabled) }

    /** Adds a user-drawn rectangle (bitmap coordinates) as a selected segment. */
    fun addManualRect(rect: Rect) {
        val photo = session.current ?: return
        val bmp = photo.workBitmap
        val clipped = Rect(rect).apply { sort() }
        // intersect() returns false (and leaves the rect alone) when the box is fully outside the photo.
        if (!clipped.intersect(0, 0, bmp.width, bmp.height)) return
        if (clipped.width() < MIN_MANUAL_SIZE || clipped.height() < MIN_MANUAL_SIZE) return
        val index = (photo.segments.maxOfOrNull { it.index } ?: -1) + 1
        photo.segments = photo.segments + Segment.manual(index, clipped)
        photo.selected.add(index)
        _state.update { it.copy(segments = photo.segments, selected = photo.selected.toSet(), manualMode = false, error = null) }
    }

    fun remove(index: Int) {
        val photo = session.current ?: return
        photo.segments = photo.segments.filterNot { it.index == index }
        photo.selected.remove(index)
        _state.update { it.copy(segments = photo.segments, selected = photo.selected.toSet()) }
    }

    /**
     * "Retake" for one item: drop that segment, stash every other selected
     * cutout from this photo, then let the caller open the camera again.
     */
    fun retake(index: Int, onReady: () -> Unit) {
        remove(index)
        viewModelScope.launch {
            stashSelected()
            onReady()
        }
    }

    private suspend fun stashSelected() {
        val photo = session.current ?: return
        val toRender = photo.segments.filter { it.index in photo.selected }
        if (toRender.isEmpty()) return
        val cutouts = withContext(container.dispatchers.default) {
            toRender.map { seg ->
                CaptureSession.Cutout(container.cutoutRenderer.render(photo.workBitmap, seg), photo.file)
            }
        }
        session.pendingCutouts.addAll(cutouts)
        photo.selected.clear()
        _state.update { it.copy(selected = emptySet(), pendingCount = session.pendingCutouts.size) }
    }

    /** Render every selected cutout, insert placeholder items, hand their ids to Confirm. */
    fun identify(onItems: (List<String>) -> Unit) {
        if (_state.value.committing) return
        _state.update { it.copy(committing = true, error = null) }
        viewModelScope.launch {
            try {
                stashSelected()
                val cutouts = session.pendingCutouts.toList()
                if (cutouts.isEmpty()) {
                    _state.update { it.copy(committing = false, error = "Select at least one item.") }
                    return@launch
                }
                val now = container.clock.nowMillis()
                val items = cutouts.map { c ->
                    ItemEntity(
                        id = Ids.newId(),
                        // Blank name marks "not identified yet"; ConfirmViewModel fills it in.
                        name = "",
                        category = "",
                        condition = Condition.GOOD,
                        searchQuery = "",
                        confidence = 0.0,
                        photoPath = c.file.absolutePath,
                        cutoutPath = c.file.absolutePath,
                        originalPhotoPath = c.originalPhoto.absolutePath,
                        createdAt = now,
                        updatedAt = now,
                    )
                }
                container.inventory.saveItems(items)
                session.clear()
                onItems(items.map { it.id })
            } catch (e: Exception) {
                _state.update { it.copy(committing = false, error = "Could not prepare cutouts: ${e.message}") }
            }
        }
    }

    companion object {
        const val MIN_MANUAL_SIZE = 24
    }
}
